src/types/index.ts

```diff
@@ -161,7 +161,12 @@ export interface RouteLocationNormalized extends _RouteLocationBase {
   matched: RouteRecordNormalized[] // non-enumerable
 }
 
-export type RouteComponent = Component
+type PublicAPIComponent = Exclude<
+  Parameters<typeof defineComponent>[0]['components'],
+  undefined
+>[string]
+
+export type RouteComponent = PublicAPIComponent
```

Comment:
I would do:
```ts
export type RouteComponent = Component | ReturnType<typeof defineComponent>
```

which will have all possible cases when using defineComponent

Reply:
I also went too complicated by looking up the Parameters of `defineComponent` 😆

Reply:
ReturnType of overload functions just use the last overload for inference, look at [this issue](https://github.com/Microsoft/TypeScript/issues/24275#issuecomment-390701982). So may we need to find another way？

Reply:
@Senasiko can you provide an overload where this is failing?

Reply:
```
export const Home = defineComponent({
  setup() {
    return {}
  },
  props: {
    test: {
      type: String,
      required: true,
    }
  },
  render() {},
});

const router = createRouter({
  history: createWebHistory(),
  routes: [{
    path: '/home/:test',
    name: 'home',
    props: true,
    component: Home,
  }],
});
```
If `defineComponent`'s param has `props` option, I will get types error in `component: Home,`, even if I write `declare type RouteComponent = Component | ReturnType<typeof defineComponent>;` in `vue-router.d.ts`

Reply:
Having the generic constructor should be enough:
```ts
declare type RouteComponent =
  | Component
  | { new (): ComponentPublicInstance }
```